### ISSUE TITLE
LSP client should be able to handle snippets that use $0 as a tabstop

### ISSUE THREAD

**Issue by eEQK** [neovim/neovim#29251]
### Problem

When LSP snippets are enabled, completion behaves in an annoying way where the snippet placeholder is not preselected.
Under the hood, this is what is being sent to the client: `textEditText = "print(${0:object})"`
![not_working](https://github.com/neovim/neovim/assets/15970653/8f1928ab-62a8-4458-881c-179414a9974b)

I've tried patching dart LSP server so that it sends `textEditText = "print(${1:object})"` instead. It works as expected:
![working](https://github.com/neovim/neovim/assets/15970653/02587f79-a6ad-477b-95e0-fa9bcbe2c271)

The first scenario works as expected in VSCode. I've checked what [LSP spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#snippet_syntax) says about it but I'd say it's pretty vague.
> The number is the order in which tab stops will be visited, whereas $0 denotes the final cursor position.

I wanted to submit this patch to the Dart SDK however after consulting it with the Dart team they said if something is unclear in LSP spec then it's generally a good idea to treat VSCode as a reference implementation of LSP client.

tl;dr I feel like it's a missing edge case that's currently not being handled by Neovim LSP

### Steps to reproduce using "nvim -u minimal_init.lua"

I've added the following test locally (`test/functional/lua/snippet_spec.lua`):

### ❌ failing test (that should pass)
```lua
  it('expands snippet with placeholder that starts counting from 0', function()
    test_expand_success({ 'print(${0:object})' }, { 'print(object)' })
    feed('1<Esc>')
    eq({ 'print(1)' }, buf_lines(0))
  end)
```
```
RUN      T12 vim.snippet expands snippet with placeholder that starts counting from 0: 7.00 ms FAIL
test/functional/lua/snippet_spec.lua:130: Expected objects to be the same.
Passed in:
(table: 0x768a2e5ee440) {
 *[1] = 'print(1object)' }
Expected:
(table: 0x768a2e69ed98) {
 *[1] = 'print(1)' }
 ```

### ✅ passing test (that should keep passing)
```lua
  it('expands snippet with placeholder that starts counting from 0', function()
    test_expand_success({ 'print(${1:object})' }, { 'print(object)' })
    feed('1<Esc>')
    eq({ 'print(1)' }, buf_lines(0))
  end)
```

### Neovim version (nvim -v)

NVIM v0.10.0

### Language server name/version

SDK 3.4.1 (I checked git blame in SDK and it's always been like that)

### Operating system/version

EndeavourOS 6.9.2-arch1-1

**Comment by clason** [comment]
>  if something is unclear in LSP spec then it's generally a good idea to treat VSCode as a reference implementation of LSP client.

**NO**. I cannot stress this enough. If something is unclear in the LSP spec then _the LSP spec should be clarified_. The correct approach is to open an issue at https://github.com/microsoft/language-server-protocol. "Do what VS Code does" is an explicit non-goal.

**Comment by DanTup** [comment]
> If something is unclear in the LSP spec then _the LSP spec should be clarified_.

I agree with this, but FWIW what I meant in my comment to the OP (on the Dart server) was that in every case where I've raised something with LSP, the clarification has been to specify VS Code's existing behaviour (which I sometimes disagree with, but is how it is).

I've opened https://github.com/microsoft/language-server-protocol/issues/1946 to ask about making the spec more explicit that $0 is also valid as a placeholder, but since neovim is already inserting the text from it, this does feel like a bug to me. If neovim wanted to treat it as only a tabstop (which could be argued that's what the spec says), then I wouldn't expect the placeholder text to have been inserted into the document.

**Comment by clason** [comment]
>  the clarification has been to specify VS Code's existing behaviour (which I sometimes disagree with, but is how it is).

Yeah, that is an issue, but at least it's explicit then (and can be argued about). 

> If neovim wanted to treat it as only a tabstop (which could be argued that's what the spec says), then I wouldn't expect the placeholder text to have been inserted into the document.

Yes, that is another possible resolution of this -- depending on how the protocol issue is resolved.

But in any case, "it's what VS Code does" simply holds no water with us -- if it's UB, our interpretation is just as valid as VS Code's, and servers need to accept that (or admit that they're actually writing a VS Code extension and not a language server).

**Comment by DanTup** [comment]
> But in any case, "it's what VS Code does" simply holds no water with us

Understood. And I actually agree with this (I contribute to the Dart LSP server, and I want it to be usable by as many clients as possible and have raised many clarification requests against LSP 🙂), I think I didn't describe this as well as I could have to the OP.

I proposed a simple text change in the PR that matches what VS Code does (which I suspect is what would ultimately be agreed) in the hope of making it easy for them to just agree and I can file a PR with that change. If for some reason it's decided that VS code's behaviour is not intended and the spec is clarified to say $0 is _only_ a tabstop and can't have placeholder text, then I'll look at changing the Dart LSP implementation.

Thanks!

**Comment by mfussenegger** [comment]
Independent of how `$0` should be treated, I'd think that `print(${1:object})$0` would make for a better snippet anyway because you get an additional jump point to leave the parenthesis. With the added bonus that it is unambiguous in regards to how `$0` behaves.

**Comment by DanTup** [comment]
@mfussenegger that's what we had originally, but doing so keeps the editor (depending on how it works) in "snippet mode" when the cursor is at $1, and sometimes the user is typing multiple lines of code as function arguments, and code completion and pressing `tab` would not behave as they expected (see https://github.com/microsoft/vscode/issues/43270 and the numerous Dart-Code issues that are linked to it).

In Flutter, it's very common to write lots of multiline code (with indenting) in arguments to constructors, so using $0 is the only way to ensure the editor isn't stuck in snippet mode while the user is trying to just type code.

**Comment by heddxh** [comment]
Now we have https://github.com/microsoft/language-server-protocol/pull/2087 merged, maybe this issue should be closed since neovim native snippet engine behavior fits the MS standard.

**Comment by intxff** [comment]
[microsoft/language-server-protocol#2087 is reverted](https://github.com/microsoft/language-server-protocol/pull/2102). This issue should be reopened.
